crates/lib/src/execution_data.rs

```diff
@@ -63,44 +126,95 @@ impl ReadoutMap {
         readout_mappings: &HashMap<String, String>,
         readout_values: &HashMap<String, ReadoutValues>,
     ) -> Self {
-        let result = readout_values
-            .iter()
-            .flat_map(|(readout_name, values)| {
-                readout_mappings
-                    .iter()
-                    .filter(|(_, program_alias)| *program_alias == readout_name)
-                    .map(|(program_name, _)| program_name.as_ref())
-                    .map(parse_readout_register)
-                    .map(|reference| Ok((reference?, values.clone())))
-                    .collect::<Result<Vec<_>, MemoryReferenceParseError>>()
-            })
-            .flatten()
-            .collect::<HashMap<_, _>>()
-            .into();
+        let mut result = ReadoutMap(HashMap::new());
+        for (readout_name, values) in readout_values {
+            readout_mappings
+                .iter()
+                .filter(|(_, program_alias)| *program_alias == readout_name)
+                .map(|(program_name, _)| program_name.as_ref())
+                .map(parse_readout_register)
```

Comment:
Could combine these into a single `filter_map`. Something like

```rust
.filter_map(|(program_name, program_alias)| {
    (*program_alias == readout_name).then(|| parse_readout_register(&program_name))
})
```

Or maybe just the first two:

```rust
.filter_map(|(program_name, program_alias)| {
    (*program_alias == readout_name).then(|| program_name.as_ref())
})
```

Reply:
Mentioning as something that I think a clippy lint would point out. Not sure if one method is better than another.

Reply:
Neat! Went with the second option, it just felt a little more clear to have `parse_readout_register` be a distinct step.